File: conn4.c
#include <limits.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define OFF_BOARD -2
#define EMPTY -1
#define LOOK_AHEAD 5
#define TABLE_SIZE 32000
#define TABLE_BIN_SIZE 10

typedef struct {
	int width;
	int height;
	int* board;
	int last_move;
	int weight;

	int refs;
} GameState;

GameState* newGameState(int width, int height) {
	GameState* toR = (GameState*) malloc(sizeof(GameState));
	toR->width = width;
	toR->height = height;
	toR->board = (int*) malloc(sizeof(int) * width * height);
	toR->weight = 0;
	toR->refs = 1;

	for (int i = 0; i < width * height; i++) {
		toR->board[i] = EMPTY;
	}

	return toR;
}

void freeGameState(GameState* gs) {
	gs->refs--;
	if (gs->refs <= 0) {
		free(gs->board);
		free(gs);
	}
}

void retainGameState(GameState* gs) {
	gs->refs++;
}

int at(GameState* gs, int x, int y) {
	if (x < 0 || y < 0)
		return OFF_BOARD;

	if (x >= gs->width || y >= gs->height)
		return OFF_BOARD;

	return gs->board[x * gs->height + y];
}

void drop(GameState* gs, int column, int player) {
	for (int i = 0; i < gs->height; i++) {
		if (at(gs, column, i) == EMPTY) {
			gs->board[column * gs->height + i] = player;
			gs->last_move = column;
			return;
		}
	}


}

int checkAt(GameState* gs, int x, int y) {
	// check across
	int found = 1;
	int curr = at(gs, x, y);
	for (int i = 0; i < 4; i++) {
		if (at(gs, x + i, y) != curr) {
			found = 0;
			break;
		}
	}

	if (found && (curr != EMPTY && curr != OFF_BOARD))
		return curr;

	// check down
	found = 1;
	for (int i = 0; i < 4; i++) {
		if (at(gs, x, y + i) != curr) {
			found = 0;
			break;
		}
	}

	if (found && (curr != EMPTY && curr != OFF_BOARD))
		return curr;

	// check diag +/+
	found = 1;
	for (int i = 0; i < 4; i++) {
		if (at(gs, x + i, y + i) != curr) {
			found = 0;
			break;
		}
	}

	if (found && (curr != EMPTY && curr != OFF_BOARD))
		return curr;

	// check diag -/+
	found = 1;
	for (int i = 0; i < 4; i++) {
		if (at(gs, x - i, y + i) != curr) {
			found = 0;
			break;
		}
	}

	if (found && (curr != EMPTY && curr != OFF_BOARD))
		return curr;

	return 0;
}

int getIncrementForArray(int* arr, int player) {
	int toR = 0;
	for (int i = 0; i < 4; i++) {
		if (arr[i] == player) {
			toR = 1;
			continue;
		}

		if (arr[i] != player && arr[i] != EMPTY) {
			return 0;
		}
	}

	return toR;
}

int countAt(GameState* gs, int x, int y, int player) {

	// check across
	int found = 0;
	int buf[4];

	for (int i = 0; i < 4; i++) {
		buf[i] = at(gs, x + i, y);
	}
    
	found += getIncrementForArray(buf, player);

	// check down
	for (int i = 0; i < 4; i++) {
		buf[i] = at(gs, x, y + i);
	}
	found += getIncrementForArray(buf, player);

	// check diag +/+
	for (int i = 0; i < 4; i++) {
		buf[i] = at(gs, x + i, y + i);
	}
	found += getIncrementForArray(buf, player);

	// check diag -/+
	for (int i = 0; i < 4; i++) {
		buf[i] = at(gs, x - i, y + i);
	}
	found += getIncrementForArray(buf, player);

	return found;
}

int getWinner(GameState* gs) {
	for (int x = 0; x < gs->width; x++) {
		for (int y = 0; y < gs->height; y++) {
			int res = checkAt(gs, x, y);
			if (res)
				return res;
		}
	}

	return 0;
}

int isDraw(GameState* gs) {
	for (int x = 0; x < gs->width; x++) {
		for (int y = 0; y < gs->height; y++) {
			if (at(gs, x, y) == EMPTY)
				return 0;
		}
	}

	return 1;
}

int getHeuristic(GameState* gs, int player, int other_player) {
	int toR = 0;
	for (int x = 0; x < gs->width; x++) {
		for (int y = 0; y < gs->height; y++) {
			toR += countAt(gs, x, y, player);
			toR -= countAt(gs, x, y, other_player);
		}
	}
	
	return toR;
}

int canMove(GameState* gs, int column) {
	for (int y = 0; y < gs->height; y++) {
		if (at(gs, column, y) == EMPTY)
			return 1;
	}

	return 0;
}

GameState* stateForMove(GameState* orig, int column, int player) {
	GameState* toR = newGameState(orig->width, orig->height);
	memcpy(toR->board, orig->board, sizeof(int) * orig->width * orig->height);
	drop(toR, column, player);
	return toR;
}

void printGameState(GameState* gs) {
	for (int i = 0; i < gs->width; i++) {
		printf("%d ", i);
	}

	printf("\n");
	
	for (int y = gs->height - 1; y >= 0 ; y--) {
		for (int x = 0; x < gs->width; x++) {
			int toP = at(gs, x, y);
			if (toP == EMPTY) {
				printf("  ");
			} else {
				printf("%d ", toP);
			}
		}
		printf("\n");
	}

	for (int i = 0; i < gs->width; i++) {
		printf("%d ", i);
	}

	printf("\n\n");
}

unsigned long hashGameState(GameState* gs) {
	unsigned long long hash = 14695981039346656037Lu;
	for (int i = 0; i < gs->width * gs->height; i++) {
		hash ^= gs->board[i];
		hash *= 1099511628211Lu;
	}
	return hash;
}

int isGameStateEqual(GameState* gs1, GameState* gs2) {
	if (gs1->width != gs2->width || gs1->height != gs2->height)
		return 0;

	for (int i = 0; i < gs1->width * gs1->height; i++) {
		if (gs1->board[i] == gs2->board[i])
			continue;
		return 0;
	}

	return 1;
}

typedef struct {
	GameState*** bins;
} TranspositionTable;


TranspositionTable* newTable() {
	TranspositionTable* toR = (TranspositionTable*) malloc(sizeof(TranspositionTable));
	toR->bins = (GameState***) malloc(sizeof(GameState**) * TABLE_SIZE);
	for (int i = 0; i < TABLE_SIZE; i++) {
		toR->bins[i] = (GameState**) malloc(sizeof(GameState*) * TABLE_BIN_SIZE);
		for (int j = 0; j < TABLE_BIN_SIZE; j++) {
			toR->bins[i][j] = NULL;
		}
	}

	return toR;
}

GameState* lookupInTable(TranspositionTable* t, GameState* k) {
	int hv = hashGameState(k) % TABLE_SIZE;
	GameState** bin = t->bins[hv];
	for (int i = 0; i < TABLE_BIN_SIZE; i++) {
		if (bin[i] == NULL) {
			return NULL;
		}

		if (isGameStateEqual(k, bin[i])) {
			return bin[i];
		}
	}
	return NULL;
}

void addToTable(TranspositionTable* t, GameState* k) {
	int hv = hashGameState(k) % TABLE_SIZE;
	GameState** bin = t->bins[hv];
	for (int i = 0; i < TABLE_BIN_SIZE; i++) {
		if (bin[i] == NULL) {
			bin[i] = k;
			retainGameState(k);
			return;
		}
	}

	fprintf(stderr, "Overflow in hash bin %d, won't store GameState\n", hv);
}

void freeTranspositionTable(TranspositionTable* t) {
	for (int i = 0; i < TABLE_SIZE; i++) {
		for (int j = 0; j < TABLE_BIN_SIZE; j++) {
			if (t->bins[i][j] != NULL)
				freeGameState(t->bins[i][j]);
		}
		free(t->bins[i]);
	}
	free(t->bins);

	free(t);
}

typedef struct {
	GameState* gs;
	int player;
	int other_player;
	int turn;
	
	int alpha;
	int beta;

	int best_move;

	TranspositionTable* ht;
} GameTreeNode;

GameTreeNode* newGameTreeNode(GameState* gs, int player, int other, int turn, int alpha, int beta, TranspositionTable* ht) {
	GameTreeNode* toR = (GameTreeNode*) malloc(sizeof(GameTreeNode));
	toR->gs = gs;
	toR->player = player;
	toR->other_player = other;
	toR->turn = turn;
	toR->alpha = alpha;
	toR->beta = beta;
	toR->best_move = -1;
	toR->ht = ht;
	return toR;
}

int heuristicForState(GameState* gs, int player, int other) {
	if (isDraw(gs))
		return 0;

	int term_stat = getWinner(gs);
	if (term_stat == player)
		return 1000;

	if (term_stat)
		return -1000;

	
	return getHeuristic(gs, player, other);
		     
}


// using a global instead of qsort_r because of emscripten support

GameTreeNode* g_node = NULL;


int ascComp(const void* a, const void* b) {
	GameTreeNode* node = g_node;
	return heuristicForState(*(GameState**) a, node->player, node->other_player) -
		heuristicForState(*(GameState**) b, node->player, node->other_player);
	
}

int desComp(const void* a, const void* b) {
	GameTreeNode* node = g_node;
	return heuristicForState(*(GameState**) b, node->player, node->other_player) -
		heuristicForState(*(GameState**) a, node->player, node->other_player);
	
}

int getWeight(GameTreeNode* node, int movesLeft) {
	int toR;
	if (getWinner(node->gs) || isDraw(node->gs) || movesLeft == 0)
		return heuristicForState(node->gs, node->player, node->other_player);

	GameState** possibleMoves = (GameState**) malloc(sizeof(GameState*) * node->gs->width);
	int validMoves = 0;
	for (int possibleMove = 0; possibleMove < node->gs->width; possibleMove++) {
		if (!canMove(node->gs, possibleMove)) {
			continue;
		}
		
		possibleMoves[validMoves] = stateForMove(node->gs, possibleMove, (node->turn ? node->player : node->other_player));
		validMoves++;
	}

	// order possibleMoves by the heuristic
	g_node = node;
	if (node->turn) {
		// qsort_r is apparently non-standard, and won't work with emscripten. So we'll need to use a global.
		qsort(possibleMoves, validMoves, sizeof(GameState*), ascComp); 
	} else {
		qsort(possibleMoves, validMoves, sizeof(GameState*), desComp); 
	}

	int best_weight = (node->turn ? INT_MIN : INT_MAX);

	for (int move = 0; move < validMoves; move++) {
		// see if the game state is already in the hash table
		GameState* inTable = lookupInTable(node->ht, possibleMoves[move]);
		int child_weight;
		int child_last_move;
		if (inTable != NULL) {
			child_weight = inTable->weight;
			child_last_move = possibleMoves[move]->last_move;

		} else {
			GameTreeNode* child = newGameTreeNode(possibleMoves[move], node->player, node->other_player, !(node->turn),
							      node->alpha, node->beta, node->ht);
			child_weight = getWeight(child, movesLeft - 1);
			child_last_move = child->gs->last_move;
			free(child);
		}

	

		possibleMoves[move]->weight = child_weight;
		addToTable(node->ht, possibleMoves[move]);

		if (movesLeft == LOOK_AHEAD)
			printf("Move %d has weight %d\n", child_last_move, child_weight);

		// alpha-beta pruning
		if (!node->turn) {
			// min node
			if (child_weight <= node->alpha) {
				// MAX ensures we will never go here
				toR = child_weight;
				goto done;
			}
			node->beta = (node->beta < child_weight ? node->beta : child_weight);
		} else {
			// max node
			if (child_weight >= node->beta) {
				// MIN ensures we will never go here
				toR = child_weight;
				goto done;
			}
			node->alpha = (node->alpha > child_weight ? node->alpha : child_weight);
		}

		if (!(node->turn)) {
			// min node
			if (best_weight > child_weight) {
				best_weight = child_weight;
				node->best_move = child_last_move;
			}
		} else {
			// max node
			if (best_weight < child_weight) {
				best_weight = child_weight;
				node->best_move = child_last_move;
			}
		}

		
	}
	toR = best_weight;
done:
	for (int i = 0; i < validMoves; i++) {
		freeGameState(possibleMoves[i]);
	}

	free(possibleMoves);
	return toR;
}

int getBestMove(GameTreeNode* node, int movesLeft) {
	getWeight(node, movesLeft);
	return node->best_move;
}


// END OF API

void checkWin(GameState* gs) {
	int win = getWinner(gs);
	
	if (win) {
		printf("Game over! %d wins!\n", win);
		printGameState(gs);
		exit(0);
	}

	if (isDraw(gs)) {
		printf("Game over! Draw!\n");
		printGameState(gs);
		exit(0);
	}
	     
}

int bestMoveForState(GameState* gs, int player, int other_player, int look_ahead) {
	TranspositionTable* t1 = newTable();
	GameTreeNode* n = newGameTreeNode(gs, player, other_player, 1, INT_MIN, INT_MAX, t1);
	int move = getBestMove(n, look_ahead);
	free(n);
	freeTranspositionTable(t1);
	return move;
}

// a couple of ease-of-use functions that will run a game in global state

GameState* globalState;

void startNewGame() {
	globalState = newGameState(7, 6);
}

void playerMove(int move) {
	drop(globalState, move, 1);
}

void computerMove(int look_ahead) {
	int move = bestMoveForState(globalState, 2, 1, look_ahead);
	drop(globalState, move, 2);
}

int isGameWon() {
	return getWinner(globalState);
}

int isGameDraw() {
	return isDraw(globalState);
}

int isEmpty(int x, int y) {
	return at(globalState, x, y) == EMPTY;
}

int pieceAt(int x, int y) {
	return at(globalState, x, y);
}


int main(int argc, char** argv) {
	startNewGame();
	while (1) {
		int move;
		printf("Move? ");
		scanf("%d", &move);
	       
		playerMove(move);
 
		printGameState(globalState);
 
		checkWin(globalState);
 
		computerMove(LOOK_AHEAD);
 
		printGameState(globalState);
 
		checkWin(globalState);
	}
 
	freeGameState(globalState);
 
	return 0;
}

File: conn4.py
class GameState:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.columns = [None for i in range(0, self.width * self.height)]

    def drop(self, column, player):
        for i in range(self.height):
            if self.__at(column, i) == None:
                break
        self.columns[(column * self.height) + i] = player
            

    def __at(self, x, y):
        if x >= self.width or x < 0:
            return "OFF"

        if y >= self.height or y < 0:
            return "OFF"

        return self.columns[(x * self.height) + y]

    def __check_at(self, x, y):
        def collapse(l):
            f = next(l)
            for i in l:
                if i != f:
                    return None
            return f

        possible = [[(x, y), (x + 1, y), (x + 2, y), (x + 3, y)],
                    [(x, y), (x, y + 1), (x, y + 2), (x, y + 3)],
                    [(x, y), (x + 1, y + 1), (x + 2, y + 2), (x + 3, y + 3)],
                    [(x, y), (x - 1, y + 1), (x - 2, y + 2), (x - 3, y + 3)]]
        possible = map(lambda possibility: map(lambda pos: self.__at(pos[0], pos[1]), possibility), possible)
        possible = map(collapse, possible)
        for res in possible:
            if res:
                return res
        return None

    def __count_at(self, x, y, for_player):
        possible = [[(x, y), (x + 1, y), (x + 2, y), (x + 3, y)],
                    [(x, y), (x, y + 1), (x, y + 2), (x, y + 3)],
                    [(x, y), (x + 1, y + 1), (x + 2, y + 2), (x + 3, y + 3)],
                    [(x, y), (x - 1, y + 1), (x - 2, y + 2), (x - 3, y + 3)]]

        possible = map(lambda possibility: map(lambda pos: self.__at(pos[0], pos[1]), possibility), possible)

        c = 0
        for p in possible:
            counts = True
            for player in p:
                if player == "OFF":
                    counts = False
                    break
                if player != None and player != for_player:
                    counts = False
                    break
            
            if counts:
                c += 1


        return c

    def __is_column_empty(self, x):
        for i in range(self.height):
            if self.__at(x, i) == None:
                return True

        return False

    def get_winner(self):
        for i in range(self.width):
            for j in range(self.height):
                res = self.__check_at(i, j)
                if res:
                    return res

        return None

    def get_heuristic(self, player, other_player):
        c = 0
        for i in range(self.width):
            for j in range(self.height):
                c += self.__count_at(i, j, player)
                c -= self.__count_at(i, j, other_player)

        return c

    def is_draw(self):
        for i in range(self.width):
            for j in range(self.height):
                if not self.__at(i, j):
                    return False
        return True

    def get_moves(self):
        moves = range(self.width)
        moves = filter(lambda x: self.__is_column_empty(x), moves)
        return moves

    def state_for_move(self, move, player):
        toR = GameState(self.width, self.height)
        toR.columns = list(self.columns)

        toR.drop(move, player)
        return toR

    def get_num_turns(self):
        return sum(map(lambda x: len(x), self.columns))

    def __str__(self):
        toR = ""
        for x in range(self.width):
            toR += str(x) + " "

        toR += "\n"
        for y in range(self.height - 1, -1, -1):
            for x in range(self.width):
                at = self.__at(x, y)
                toR += (at if at else " ") + " "
            toR += "\n"

        for x in range(self.width):
            toR += str(x) + " "
        toR += "\n"

        return toR

    def __hash__(self):
        c = 0
        first_found = None
        for i in range(self.width):
            for j in range(self.height):
                here = self.__at(i, j)
                if here != None and first_found == None:
                    first_found = here
                if here == first_found:
                    c += 3**((i * self.height)+j)
                elif here != None and here != first_found:
                    c += 2 * 3**((i * self.height) + j)

        return c

prune = 0
no_prune = 0
terminals = 0
table_hits = 0
class GameTreeNode:
    def __init__(self, game_state, player, other_player, turn, alpha, beta, transpose, p):
        self.state = game_state
        self.player = player
        self.other_player = other_player
        self.turn = turn
        self.weight = 0
        self.transposition_table = transpose

        self.alpha = alpha
        self.beta = beta

        self.min_child = None
        self.max_child = None
        self.p = p

    def __is_min_node(self):
        return not self.turn

    def __is_max_node(self):
        return self.turn

    def is_terminal(self):
        return self.state.get_winner() or self.state.is_draw()

    def __heuristic_for_state(self, state):
        term_stat = state.get_winner() or state.is_draw()
        if term_stat:
            if term_stat and term_stat == self.player:
                return 1000
            if term_stat and term_stat == True:
                return 0
            return -1000

        return state.get_heuristic(self.player, self.other_player)

    def get_weight(self, max_moves=1000):
        global prune, no_prune, table_hits, terminals
        print(prune, no_prune, table_hits, terminals)

      

        term_stat = self.is_terminal()
        if term_stat:
            terminals += 1
            if term_stat and term_stat == self.player:
                return 1000
            if term_stat and term_stat == True:
                return 0

            return -1000

        if max_moves == 0:
            # use the heurisitic
            return self.state.get_heuristic(self.player, self.other_player)


        move_states = map(lambda x: (x, self.state.state_for_move(x, self.player if self.turn else self.other_player)), self.state.get_moves())
        new_states = sorted(move_states, key=lambda x: self.__heuristic_for_state(x[1]), reverse=self.__is_max_node())
        for move, state in new_states:
            child = GameTreeNode(state, self.player, self.other_player, not self.turn, 
                                 self.alpha, self.beta, self.transposition_table, False)

            if hash(child.state) not in self.transposition_table:
                child = (child.get_weight(max_moves=max_moves - 1), child)
                self.transposition_table[hash(child[1].state)] = child[0]
            else:
                child = (self.transposition_table[hash(child.state)], child)
                table_hits += 1

            if self.p:
                print("Move", move, "has weight", child[0])

            # alpha-beta pruning
            if self.__is_min_node():
                if child[0] <= self.alpha:
                    # stop search. MAX will ensure we never get here.
                    prune += 1
                    return child[0]
                self.beta = min(self.beta, child[0])
                
                
            if self.__is_max_node():
                if child[0] >= self.beta:
                    # stop search. MIN will ensure we never get here.
                    prune += 1
                    return child[0]
                self.alpha = max(self.alpha, child[0])
             
            no_prune += 1 


            if self.min_child == None:
                self.min_child = child
            else:
                self.min_child = min(self.min_child, child, key=lambda x: x[0])

            if self.max_child == None:
                self.max_child = child
            else:
                self.max_child = max(self.max_child, child, key=lambda x: x[0])

        if self.__is_max_node():
            return self.max_child[0] if self.max_child else None
        else:
            return self.min_child[0] if self.min_child else None

    def get_best_child(self, max_moves=1000):
        self.get_weight(max_moves=max_moves)
        if self.__is_min_node():
            return self.min_child[1]

        return self.max_child[1]

            


def check_win(g):
    win = g.get_winner()
    if win:
        print("Game over!", win, "wins!")
        print(g)
        exit(0)

    if g.is_draw():
        print("Game over!", "It's a draw!")
        print(g)
        exit(0)

def human_move(g):
    print(g)
    move = int(input("Select a column to drop a piece in:"))
    g = g.state_for_move(move, "h")
    check_win(g)
    return g

def computer_move(g, player="c", other="h", moves=5):
    gtn = GameTreeNode(g, player, other, True, float("-inf"), float("inf"), dict(), True)
    gtn = gtn.get_best_child(max_moves=moves)
    g = gtn.state
    check_win(g)
    return g


def hvc():
    g = GameState(7, 6)
    first = input("Who should go first? (h|c): ")
    if first == "h":
        while True:
            g = human_move(g)
            g = computer_move(g)

    if first == "c":
        while True:
            g = computer_move(g)
            g = human_move(g)

def cvc():
    g = GameState(4, 4)
    while True:
        g = computer_move(g, player="1", other="2", moves=7)
        print(g)
        g = computer_move(g, player="2", other="1", moves=5)
        print(g)

cvc()

